fhir-mapping/src/test/ml-modules/root/test/suites/testUtils.sjs

```diff
@@ -0,0 +1,241 @@
+   * @return {void}
+   */
+  insertOriginalDocumentCollections(documents, database) {
+    xdmp.invokeFunction(
+      () => { documents.forEach(d => { d.collections = xdmp.documentGetCollections(d.baseURI); }); },
+      { database },
+    );
+  },
+
+  /**
+   * Add the given documents' original permissions from the source database to the objects containing them in order to be inserted into the destination database
+   *
+   * @param  {DocumentWrapper[]} documents The documents to get permissions for
+   * @param  {number}            database  The source database ID
+   *
+   * @return {void}
+   */
+  insertOriginalDocumentPermissions(documents, database) {
+    xdmp.invokeFunction(
+      () => { documents.forEach(d => { d.permissions = xdmp.documentGetPermissions(d.baseURI); }); },
+      { database },
+    );
+  },
+
+  /**
+   * Wrapper to attempt a callback and log any errors that occurred. DOES NOT RE-THROW ERRORS
+   *
+   * @param  {(...any[]) => T} cb   A function callback to be called
+   * @param  {...any[]}        args Arguments to pass to the callback at runtime
+   *
+   * @return {T}
+   */
+  attempt(cb, ...args) {
+    try {
+      return cb(...args);
+    } catch (e) {
+      this.logger.error('%s:\n%s', e.message, e.stack);
+    }
+  },
+  /**
+   * Flatten an array by one level, e.g.; [[1], [2, [3]], 4] -> [1, 2, [3], 4]
+   *
+   * @param  {any[]} arr The array to flatten
+   *
+   * @return {any[]}
+   */
+  flatten(arr) {
+    // NOTE: Flatten the array using Array.prototype.reduce (Array.prototype.flat is not supported yet)
+    return arr.reduce((acc, res) => acc.concat(res), []);
+  },
+  /**
+   * Flatten an array after having mapped its values using the given callback function
+   *
+   * @param  {any[]}        arr The array to map/flatten
+   * @param  {(any) => any} cb  The mapping function callback
+   *
+   * @return {any[]}
+   */
+  flatMap(arr, cb) {
+    return this.flatten(arr.map(cb));
+  },
+  /**
+   * Require the given object to have values for the given property keys
+   *
+   * @param  {Object}      obj    The object to check
+   * @param  {...string[]} props  The properties to require
+   *
+   * @return {void}
+   *
+   * @throws If the object is missing any of the required properties, or the property is null/undefined
+   */
+  requireProperties(obj, ...props) {
+    const keys = Object.keys(obj);
+    const missing = props.filter(prop => !keys.includes(prop) || obj[prop] === undefined || obj[prop] === null);
+
+    if (missing.length > 0) {
+      throw new Error(`Missing required properties in provided object: ${missing.map(k => `"${k}"`).join(', ')}`);
+    }
+  },
+
+  /**
+   * Load test documents along the given path from the given source database into the given destination database. Optional: load recursively along path, override collections, permissions on inserted documents.
+   *
+   * @param  {LoadOpts} opts The options
+   *
+   * @return {void}
+   */
+  loadTestDocuments(opts) {
```

Comment:
This looks like excellent code, but I'm worried it conflicts with or replaces the ml-unit-test loadTestFile() function. Not sure if there are gaps this fills, and if there are if the ml-unit-test project should get an upgraded data loading function (maybe this one) vs adding it here.

Reply:
I added this (and the associated `unloadTestDocuments` below) because I found ml-unit-test's `loadTestFile` to be somewhat poorly documented, and a bit restrictive in scope. To the best of my knowledge it shouldn't conflict with the existing functions at all. I did consider opening a pull request in the ml-unit-test repository to introduce these functions but since their entire library is written in XQuery and I'm still generally unfamiliar with that, I haven't done so.